### PR TITLE
Switch the application tree to async structure tree

### LINKED ISSUE #124: Explorer: A spinner (progress indicator) should inform that the unfolded node is loading it's children
[kind/Enhancement]

## Which functionality do you think we should update/improve?
If you unfold ex. a project node in the explorer, you notice a slight delay until the project lists its applications. This delay is caused by odo loading the applications for the project. While the loading is in progress, nothing happens, the tree seems frozen. Even though this is not a very long wait (depending on your network- and cluster-speed) this inactivity is confusing/annoying. IntelliJ encourages using progress indicators for tree nodes . when loading happens so that the user has immediate feedback: https://jetbrains.github.io/ui/controls/progress_indicators/#tree-list-tool-window
We should show a Loading node with a spinner (or display a spinner in the parent node) while the node is loading it's children.
image
btw. vscode-openshift-connector does exactly this, it is displaying a spinner while the node is loading it's children.

### PR DESCRIPTION
**Pull request by jeffmaury**
Fixes #124

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

## Was the change discussed in an issue?
fixes #???

### CODE REVIEW

**Comment by sbouchet** [comment]
i can see a lot of errors when trying locally on my dev env:
```
2021-01-29 11:54:45,806 [ 221023]  ERROR - rains.concurrency.AsyncPromise - IntelliJ IDEA 2020.3  Build #IU-203.5981.155 
2021-01-29 11:54:45,806 [ 221023]  ERROR - rains.concurrency.AsyncPromise - JDK: 11.0.9; VM: OpenJDK 64-Bit Server VM; Vendor: JetBrains s.r.o. 
2021-01-29 11:54:45,807 [ 221024]  ERROR - rains.concurrency.AsyncPromise - OS: Linux 
2021-01-29 11:54:45,807 [ 221024]  ERROR - rains.concurrency.AsyncPromise - Last Action: org.jboss.tools.intellij.openshift.actions.cluster.SetCurrentClusterAction 
2021-01-29 11:54:45,808 [ 221025]  ERROR - rains.concurrency.AsyncPromise - Argument for @NotNull parameter 'descriptor' of com/intellij/ui/tree/StructureTreeModel$Node.<init> must not be null 
java.lang.IllegalArgumentException: Argument for @NotNull parameter 'descriptor' of com/intellij/ui/tree/StructureTreeModel$Node.<init> must not be null
	at com.intellij.ui.tree.StructureTreeModel$Node.$$$reportNull$$$0(StructureTreeModel.java)
	at com.intellij.ui.tree.StructureTreeModel$Node.<init>(StructureTreeModel.java)
	at com.intellij.ui.tree.StructureTreeModel$Node.<init>(StructureTreeModel.java:427)
	at com.intellij.ui.tree.StructureTreeModel.getValidChildren(StructureTreeModel.java:389)
	at com.intellij.ui.tree.StructureTreeModel.validateChildren(StructureTreeModel.java:298)
	at com.intellij.ui.tree.StructureTreeModel.getNode(StructureTreeModel.java:292)
	at com.intellij.ui.tree.StructureTreeModel.getChildren(StructureTreeModel.java:312)
	at com.intellij.ui.tree.AsyncTreeModel$CmdGetChildren.getNode(AsyncTreeModel.java:561)
	at com.intellij.ui.tree.AsyncTreeModel$Command.get(AsyncTreeModel.java:456)
	at com.intellij.ui.tree.AsyncTreeModel$Command.get(AsyncTreeModel.java:422)
	at com.intellij.util.concurrency.Invoker$Task.run(Invoker.java:316)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1137)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.lambda$runInReadActionWithWriteActionPriority$0(ProgressIndicatorUtils.java:76)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.runActionAndCancelBeforeWrite(ProgressIndicatorUtils.java:152)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.lambda$runWithWriteActionPriority$1(ProgressIndicatorUtils.java:113)
	at com.intellij.openapi.progress.ProgressManager.lambda$runProcess$0(ProgressManager.java:59)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:178)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:658)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:610)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:65)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:165)
	at com.intellij.openapi.progress.ProgressManager.runProcess(ProgressManager.java:59)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.runWithWriteActionPriority(ProgressIndicatorUtils.java:110)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.runInReadActionWithWriteActionPriority(ProgressIndicatorUtils.java:76)
	at com.intellij.util.concurrency.Invoker.invokeSafely(Invoker.java:205)
	at com.intellij.util.concurrency.Invoker.lambda$offerSafely$0(Invoker.java:183)
	at com.intellij.util.concurrency.Invoker$Background.lambda$offer$0(Invoker.java:541)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at com.intellij.util.concurrency.SchedulingWrapper$MyScheduledFutureTask.run(SchedulingWrapper.java:220)
	at com.intellij.util.concurrency.BoundedTaskExecutor.doRun(BoundedTaskExecutor.java:216)
	at com.intellij.util.concurrency.BoundedTaskExecutor.access$200(BoundedTaskExecutor.java:27)
	at com.intellij.util.concurrency.BoundedTaskExecutor$1.execute(BoundedTaskExecutor.java:195)
	at com.intellij.util.ConcurrencyUtil.runUnderThreadName(ConcurrencyUtil.java:208)
	at com.intellij.util.concurrency.BoundedTaskExecutor$1.run(BoundedTaskExecutor.java:184)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:668)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:665)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:665)
	at java.base/java.lang.Thread.run(Thread.java:834)
2021-01-29 11:54:45,809 [ 221026]  ERROR - rains.concurrency.AsyncPromise - IntelliJ IDEA 2020.3  Build #IU-203.5981.155 
2021-01-29 11:54:45,809 [ 221026]  ERROR - rains.concurrency.AsyncPromise - JDK: 11.0.9; VM: OpenJDK 64-Bit Server VM; Vendor: JetBrains s.r.o. 
2021-01-29 11:54:45,809 [ 221026]  ERROR - rains.concurrency.AsyncPromise - OS: Linux 
2021-01-29 11:54:45,810 [ 221027]  ERROR - rains.concurrency.AsyncPromise - Last Action: org.jboss.tools.intellij.openshift.actions.cluster.SetCurrentClusterAction 
```

**Comment by sbouchet** [comment]
please also fix create project and delete project  actions. it does not refresh the tree when selected.

**Comment by sbouchet** [comment]
still experiencing refresh issues. below the step by step: 
starting with remote 3.11 cluster and 1 project ok.
create project 'p1' ->  not displayed.
create another project 'p2' -> cluster node refresh and shows only 2 projects ( initial and 'p1' ) 
delete 'p1' -> cluster node refresh  and shows 2 projects ( initial and 'p2' ) 

debugging to try to understand the event workflow.

**Comment by jeffmaury** [comment]
> 
> 
> still experiencing refresh issues. below the step by step:
> starting with remote 3.11 cluster and 1 project ok.
> create project 'p1' -> not displayed.
> create another project 'p2' -> cluster node refresh and shows only 2 projects ( initial and 'p1' )
> delete 'p1' -> cluster node refresh and shows 2 projects ( initial and 'p2' )
> 
> debugging to try to understand the event workflow.

Managed to reproduce. This is an odo issue (https://github.com/openshift/odo/issues/4426)
So I think we're good to go as this is non blocker (refresh). Probably needs to be documented in README before the release

**Comment by sbouchet** [comment]
> Managed to reproduce. This is an odo issue ([openshift/odo#4426](https://github.com/openshift/odo/issues/4426))
> So I think we're good to go as this is non blocker (refresh). Probably needs to be documented in README before the release

ok. my debug also  shows the refresh event ask odo to get the project list and get a wrong project list. 
posing thread/waiting before getting the list works and returns the correct project list. 

so I agree to put a warning on readme.

File: README.md [hunk 1]
```diff
@@ -35,7 +35,15 @@ For more information about devfiles, see the [devfile docs](https://docs.devfile
...
 ![](images/0.4.0/openshift1.gif)
 
-## WARNING !!! Breaking Changes
+## WARNING
+
+### Known issues
+
+When creating a new project, it is possible that the newly created project will not immediately
+appear in the Application Explorer tree. This is caused by a synchronization issue in the underlying
+CLI tool odo. A workaround is to refresh the tree.
```
Comment (sbouchet) [inline README.md:44]: Maybe worth linking to the underlying odo issue? also, how do user refresh the tree manually? there is no such action right now.

**Review by sbouchet — APPROVED** [review]
LGTM.

**Pull request merged**